chalk-solve/src/recursive.rs

```diff
@@ -26,6 +33,104 @@ pub(crate) struct RecursiveContext<I: Interner> {
     caching_enabled: bool,
 }
 
+pub(crate) struct RecursiveInferenceTableImpl<I: Interner> {
+    pub(crate) infer: InferenceTable<I>,
+}
+
+impl<I: Interner> RecursiveInferenceTable<I> for RecursiveInferenceTableImpl<I> {
```

Comment:
I started to wonder if maybe it's worth having this notion of an `InferenceTable` (as a trait) in `chalk_ir`. A lot of these functions are also used for slg. Notable differences include `invert_then_canonicalize` and actually using the `free_vars` from `canonicalize`

Reply:
I think the inference code *itself* should maybe move to chalk-ir, actually, but we should discuss, especially in light of your idea to "invert" the dependency layers.